cli/tests/test_fix_command.rs

```diff
@@ -0,0 +1,513 @@
+    Added 0 files, modified 1 files, removed 0 files
+    "###);
+    let content = test_env.jj_cmd_success(&repo_path, &["print", "file", "-r", "a"]);
+    insta::assert_snapshot!(content, @"oof\n");
+    let content = test_env.jj_cmd_success(&repo_path, &["print", "file", "-r", "b"]);
+    insta::assert_snapshot!(content, @"rab\n");
+    let content = test_env.jj_cmd_success(&repo_path, &["print", "file", "-r", "c"]);
+    insta::assert_snapshot!(content, @"rab\n");
+    let content = test_env.jj_cmd_success(&repo_path, &["print", "file", "-r", "d"]);
+    insta::assert_snapshot!(content, @"oof\n");
+
+    // Each new content string only appears once in the log, because all the other
+    // inputs (like file name) were identical, and so the results were re-used. We
+    // sort the log because the order of execution inside `jj fix` is undefined.
+    insta::assert_snapshot!(sorted_lines(repo_path.join("file-fixlog")), @"oof\nrab\n");
+}
+
+fn sorted_lines(path: PathBuf) -> String {
+    let mut log: Vec<_> = std::fs::read_to_string(path.as_os_str())
+        .unwrap()
+        .lines()
+        .map(String::from)
+        .collect();
+    log.sort();
+    log.join("\n")
+}
+
+#[test]
+fn test_no_op() {
+    let (test_env, repo_path) = init_with_fake_formatter("");
+    std::fs::write(repo_path.join("file"), "content").unwrap();
+
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["fix", "-s", "@"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Fixed 1 commits of 1 checked.
+    Working copy now at: qpvuntsm 38946f90 (no description set)
+    Parent commit      : zzzzzzzz 00000000 (empty) (no description set)
+    Added 0 files, modified 1 files, removed 0 files
+    "###);
+    let content = test_env.jj_cmd_success(&repo_path, &["print", "file", "-r", "@"]);
+    insta::assert_snapshot!(content, @"content");
+}
+
+#[test]
+fn test_failure() {
+    let (test_env, repo_path) = init_with_fake_formatter("--fail");
+    std::fs::write(repo_path.join("file"), "content").unwrap();
+
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["fix", "-s", "@"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    Fixed 0 commits of 1 checked.
+    Nothing changed.
+    "###);
+    let content = test_env.jj_cmd_success(&repo_path, &["print", "file", "-r", "@"]);
+    insta::assert_snapshot!(content, @"content");
+}
+
+#[test]
+fn test_stderr_success() {
+    let (test_env, repo_path) = init_with_fake_formatter("--stderr 'error' --stdout 'new content'");
+    std::fs::write(repo_path.join("file"), "old content").unwrap();
+
+    // TODO: Associate the stderr lines with the relevant tool/file/commit instead
+    // of passing it through directly.
+    let (stdout, stderr) = test_env.jj_cmd_ok(&repo_path, &["fix", "-s", "@"]);
+    insta::assert_snapshot!(stdout, @"");
+    insta::assert_snapshot!(stderr, @r###"
+    errorFixed 1 commits of 1 checked.
```

Comment:
nit: we can probably expect the formatter to include a trailing newline, so let's do that in the fake error message too?

Reply:
Actually, I think we will want the tests to cover stderr output from tools without newlines. Until we pipe stderr to format it with commit/file context info, we can't add our own newlines to keep it neat. And I don't think we can pass newlines in the argument list here, so we would have to add some kind of unescaping to the fake formatter.

I would just leave this for now as a reminder to cover this case in the future.